cvat-ui/src/utils/redux.ts

```diff
@@ -0,0 +1,18 @@
+import { Action, ActionCreatorsMapObject, AnyAction } from 'redux';
+import { ThunkAction as _ThunkAction } from 'redux-thunk';
+import { CombinedState } from '../reducers/interfaces';
+
+export interface ActionWithPayload<T, P> extends Action<T> {
+    payload: P;
+}
+
+export function createAction<T extends string>(type: T): Action<T>;
+export function createAction<T extends string, P>(type: T, payload: P): ActionWithPayload<T, P>;
+export function createAction<T extends string, P>(type: T, payload?: P) {
+    return typeof payload === 'undefined' ? { type } : { type, payload };
+}
+
```

Comment:
I am not very get on well with typescript, and I am trying to sort it out. 

They are overloads, aren't they?
I [read](https://howtodoinjava.com/typescript/function-overloading/), that all overloads must have the same return type

Reply:
Why do not extend type redux::Action by adding a payload property that has a generic type?

Something like that:
```ts
interface Action<T extends string, P>  implements  ReduxAction<T> {
    type: T,
    payload: P,
}

export function createAction<T extends string>(type: T): Action<T, undefined>;
export function createAction<T extends string, P>(type: T, payload: P): Action<T, P>;
export function createAction<T extends string, P>(type: T, payload?: P) {
    return typeof payload === 'undefined' ? { type } : { type, payload };
}

```

Reply:
It's more question, than offer.

Reply:
Oh, actually you do almost the same, but nevertheless return different types.

Reply:
Okay, I will investigate about